models/taskRequests.js

```diff
@@ -0,0 +1,176 @@
+const tasksModel = require("./tasks");
+const userModel = require("./users");
+const userStatusModel = require("./userStatus");
+
+/**
+ * Fetch all task requests
+ *
+ * @return {Object}
+ */
+const fetchTaskRequests = async () => {
+  try {
+    const taskRequestsSnapshot = await taskRequestsCollection.get();
+
+    const taskRequests = [];
+
+    const visitedUsers = new Set();
+    const userSnapshot = [];
+    const users = [];
+
+    taskRequestsSnapshot.forEach((taskRequest) => {
+      const taskRequestData = taskRequest.data();
+      const requestors = taskRequestData.requestors;
+
+      requestors.forEach((requestor) => {
+        if (!visitedUsers.has(requestor)) {
+          visitedUsers.add(requestor);
+
+          userSnapshot.push(userModel.fetchUser({ userId: requestor }));
+        }
+      });
+
+      taskRequests.push({ id: taskRequest.id, ...taskRequestData });
+    });
+
+    const usersData = await Promise.all(userSnapshot);
+    usersData.forEach((user) => {
+      users.push(user.user);
+    });
+
+    return {
+      taskRequests,
+      users,
+    };
+  } catch (err) {
+    logger.error("error fetching tasks", err);
+    throw err;
+  }
+};
+
+/**
+ * Creates a task request
+ *
+ * @param taskId { string }: id of task request
+ * @return {Promise<{taskRequest: Object}>}
+ */
+const addOrUpdate = async (taskId, userId) => {
+  try {
+    const taskRequestsSnapshot = await taskRequestsCollection.where("taskId", "==", taskId).get();
+    const [taskRequestRef] = taskRequestsSnapshot.docs;
+    const taskRequestData = taskRequestRef?.data();
+
+    const { userExists, user } = await userModel.fetchUser({ userId });
+    const { userStatusExists, data: userStatus } = await userStatusModel.getUserStatus(userId);
+
+    if (!userExists) {
+      return { userDoesNotExists: true };
+    }
+    if (!userStatusExists) {
+      return { userStatusDoesNotExist: true };
+    }
+    if (userStatus.currentStatus.state === userState.OOO) {
+      return { isUserOOO: true };
+    }
+    if (userStatus.currentStatus.state === userState.ACTIVE) {
+      return { isUserActive: true };
+    }
+    if (taskRequestData) {
+      const currentRequestors = taskRequestData.requestors;
+      const alreadyRequesting = currentRequestors.some((requestor) => requestor === user.id);
+      if (alreadyRequesting) {
+        return { alreadyRequesting };
+      }
+
+      const updatedRequestors = [...currentRequestors, user.id];
+      await taskRequestsCollection.doc(taskRequestRef.id).update({ requestors: updatedRequestors });
+
+      return {
+        isUpdate: true,
+        requestors: updatedRequestors,
+      };
+    }
+
+    const { taskData } = await tasksModel.fetchTask(taskId);
+    if (!taskData) {
+      return { taskDoesNotExist: true };
+    }
+
+    const newTaskRequest = {
+      isNoteworthy: taskData.isNoteworthy,
+      priority: taskData.priority,
+      purpose: taskData.purpose || "",
+      requestors: [userId],
```

Comment:
what if we want to get all the taksRequest for a particular user?
It will have to get all the taskRequest and then loop through the requestors array and then find these.

Reply:
We'll be using query operator `arrayContains`